app/store/keys/keys.actions.ts

```diff
@@ -1,21 +1,93 @@
 import { push } from 'connected-react-router';
 import { createAction, Dispatch } from '@reduxjs/toolkit';
+import { useHistory } from 'react-router';
+import bcryptjs from 'bcryptjs';
 import { generateMnemonicRootKeychain } from '@blockstack/keychain';
+import { encryptMnemonic, decryptMnemonic } from 'blockstack';
+
 import routes from '../../constants/routes.json';
 import { MNEMONIC_ENTROPY } from '../../constants';
+import { RootState } from '../index';
+import { persistSalt, persistEncryptedMnemonic } from '../../utils/disk-store';
+import { safeAwait } from '../../utils/safe-await';
+import log from 'electron-log';
+import { selectMnemonic } from './keys.reducer';
+
+type History = ReturnType<typeof useHistory>;
 
 export const persistMnemonicSafe = createAction<string>('keys/save-mnemonic-safe');
 
 export const persistMnemonic = createAction<string>('keys/save-mnemonic');
 
-export const removeMnemonic = createAction<string>('keys/remove-mnemonic');
-
-export const onboardingMnemonicGenerationStep = ({
-  stepDelayMs,
-}: {
-  stepDelayMs: number;
-}) => async (dispatch: Dispatch) => {
-  const key = await generateMnemonicRootKeychain(MNEMONIC_ENTROPY);
-  dispatch(persistMnemonicSafe(key.plaintextMnemonic));
-  setTimeout(() => dispatch(push(routes.SECRET_KEY)), stepDelayMs);
-};
+export const addData = createAction<any>('keys/add-all');
+
+interface SetPasswordSuccess {
+  salt: string;
+  encryptedMnemonic: string;
+}
+export const setPasswordSuccess = createAction<SetPasswordSuccess>('keys/set-password-success');
+
+export function onboardingMnemonicGenerationStep({ stepDelayMs }: { stepDelayMs: string }) {
+  return async (dispatch: Dispatch) => {
+    const key = await generateMnemonicRootKeychain(MNEMONIC_ENTROPY);
+    dispatch(persistMnemonicSafe(key.plaintextMnemonic));
+    setTimeout(() => dispatch(push(routes.SECRET_KEY)), stepDelayMs);
+  };
+}
+
+const randomSalt = bcryptjs.genSaltSync(12);
```

Comment:
Hm, I'm wondering if this is the right place to do this?

I know we discussed that having a random salt isn't totally necessary, for our purposes.

But, by putting this here, are we potentially slowing down imports by running this synchronous function? It's probably fine, I just am not sure. Maybe it would be better to memoize the salt, but generate it inside of `setPassword` if it's not yet in memory?

Reply:
memoized 👍